### PR TITLE
Add pod start and stop

### PR DESCRIPTION
**Pull request by haircommander**
As well as added tests, man pages, and completions.
Also reformat and refactor a couple of other small things in the other pod commands.

Signed-off-by: haircommander <pehunt@redhat.com>

### CODE REVIEW

File: cmd/podman/pod_rm.go [hunk 1]
```diff
@@ -38,13 +38,15 @@ var (
...
-	ctx := getContext()
-	if err := validateFlags(c, rmFlags); err != nil {
-		return err
-	}
-
-	if c.Bool("latest") && c.Bool("all") {
+	args := c.Args()
+	if (c.Bool("all") || c.Bool("latest")) && len(args) > 0 {
+		return errors.Errorf("no arguments are needed with --all or --latest")
+	} else if c.Bool("all") && c.Bool("latest") {
```
Comment (rhatdan) [inline cmd/podman/pod_rm.go]: No need for else.

File: cmd/podman/pod_rm.go [hunk 2]
```diff
@@ -38,13 +38,15 @@ var (
...
 		return errors.Errorf("--all and --latest cannot be used together")
+	} else if len(args) < 1 && !c.Bool("all") && !c.Bool("latest") {
```
Comment (rhatdan) [inline cmd/podman/pod_rm.go]: No need for else.

File: cmd/podman/pod_rm.go [hunk 3]
```diff
@@ -38,13 +38,15 @@ var (
...
+		return errors.Errorf("you must provide at least one pod name or id")
+	} else if err := validateFlags(c, startFlags); err != nil {
```
Comment (rhatdan) [inline cmd/podman/pod_rm.go]: No need for else.

File: cmd/podman/pod_rm.go [hunk 4]
```diff
@@ -53,20 +55,15 @@ func podRmCmd(c *cli.Context) error {
...
 	var delPods []*libpod.Pod
 	var lastError error
-	if c.IsSet("all") {
+	if c.Bool("all") {
 		delPods, err = runtime.GetAllPods()
 		if err != nil {
 			return errors.Wrapf(err, "unable to get pod list")
 		}
-	} else if c.IsSet("latest") {
+	} else if c.Bool("latest") {
```
Comment (rhatdan) [inline cmd/podman/pod_rm.go]: No need for else?  You checked above correct?

File: cmd/podman/pod_start.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+	"github.com/projectatomic/libpod/cmd/podman/libpodruntime"
+	"github.com/projectatomic/libpod/libpod"
+	"github.com/urfave/cli"
+)
+
+var (
+	podStartFlags = []cli.Flag{
+		cli.BoolFlag{
+			Name:  "all, a",
+			Usage: "stop all running pods",
```
Comment (rhatdan) [inline cmd/podman/pod_start.go]: Start all pods?

File: cmd/podman/pod_start.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+		ArgsUsage:              "POD-NAME [POD-NAME ...]",
+		UseShortOptionHandling: true,
+	}
+)
+
+func podStartCmd(c *cli.Context) error {
+	args := c.Args()
+	if (c.Bool("all") || c.Bool("latest")) && len(args) > 0 {
+		return errors.Errorf("no arguments are needed with --all or --latest")
+	} else if c.Bool("all") && c.Bool("latest") {
```
Comment (rhatdan) [inline cmd/podman/pod_start.go]: No need for else.

File: cmd/podman/pod_start.go [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+		return errors.Errorf("--all and --latest cannot be used together")
+	} else if len(args) < 1 && !c.Bool("all") && !c.Bool("latest") {
```
Comment (rhatdan) [inline cmd/podman/pod_start.go]: Ditto

File: cmd/podman/pod_start.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+		return errors.Errorf("you must provide at least one pod name or id")
+	} else if err := validateFlags(c, startFlags); err != nil {
```
Comment (rhatdan) [inline cmd/podman/pod_start.go]: ditto

File: cmd/podman/pod_start.go [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+	var pods []*libpod.Pod
+	var lastError error
+
+	if c.Bool("all") {
+		pods, err = runtime.Pods()
+		if err != nil {
+			return errors.Wrapf(err, "unable to get running pods")
+		}
+	} else if c.Bool("latest") {
```
Comment (rhatdan) [inline cmd/podman/pod_start.go]: No need for else.

File: cmd/podman/pod_start.go [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+		pod, err := runtime.GetLatestPod()
+		if err != nil {
+			return errors.Wrapf(err, "unable to get latest pod")
+		}
+		pods = append(pods, pod)
+	} else {
```
Comment (rhatdan) [inline cmd/podman/pod_start.go]: No need for else.

File: cmd/podman/pod_stop.go [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+	"github.com/urfave/cli"
+)
+
+var (
+	podStopFlags = []cli.Flag{
+		cli.BoolFlag{
+			Name:  "all, a",
+			Usage: "stop all running pods",
+		},
+		cli.BoolFlag{
```
Comment (rhatdan) [inline cmd/podman/pod_stop.go]: When would I not want to do this?

File: cmd/podman/pod_stop.go [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+		Action:      podStopCmd,
+		ArgsUsage:   "POD-NAME [POD-NAME ...]",
+	}
+)
+
+func podStopCmd(c *cli.Context) error {
+	args := c.Args()
+	if (c.Bool("all") || c.Bool("latest")) && len(args) > 0 {
+		return errors.Errorf("no arguments are needed with --all or --latest")
+	} else if c.Bool("all") && c.Bool("latest") {
```
Comment (rhatdan) [inline cmd/podman/pod_stop.go]: No need for else.

File: cmd/podman/pod_stop.go [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+		return errors.Errorf("--all and --latest cannot be used together")
+	} else if len(args) < 1 && !c.Bool("all") && !c.Bool("latest") {
```
Comment (rhatdan) [inline cmd/podman/pod_stop.go]: ditto

File: cmd/podman/pod_stop.go [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+		return errors.Errorf("you must provide at least one pod name or id")
+	} else if err := validateFlags(c, stopFlags); err != nil {
```
Comment (rhatdan) [inline cmd/podman/pod_stop.go]: ditto

File: cmd/podman/pod_stop.go [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+	var pods []*libpod.Pod
+	var lastError error
+
+	if c.Bool("all") {
+		pods, err = runtime.Pods()
+		if err != nil {
+			return errors.Wrapf(err, "unable to get running pods")
+		}
+	} else if c.Bool("latest") {
```
Comment (rhatdan) [inline cmd/podman/pod_stop.go]: No need for else

File: cmd/podman/pod_stop.go [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+		pod, err := runtime.GetLatestPod()
+		if err != nil {
+			return errors.Wrapf(err, "unable to get latest pod")
+		}
+		pods = append(pods, pod)
+	} else {
```
Comment (rhatdan) [inline cmd/podman/pod_stop.go]: ditto

File: cmd/podman/pod_stop.go [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+
+	for _, pod := range pods {
+		ctr_errs, err := pod.Stop(c.Bool("cleanup"))
+		if err != nil {
+			if lastError != nil {
+				fmt.Fprintln(os.Stderr, lastError)
+			}
+			lastError = errors.Wrapf(err, "unable to stop pod %q", pod.ID())
+			continue
+		} else if ctr_errs != nil {
```
Comment (rhatdan) [inline cmd/podman/pod_stop.go]: no  need for else.

File: completions/bash/podman [hunk 18]
```diff
@@ -2154,6 +2199,9 @@ _podman_pod() {
      esac
 }
 
+
+
```
Comment (rhatdan) [inline completions/bash/podman]: Why empty lines.

File: docs/podman-pod-start.1.md [hunk 19]
```diff
@@ -0,0 +1,38 @@
+% podman-pod-start "1"
+
+## NAME
+podman\-pod\-start - Start one or more containers
```
Comment (rhatdan) [inline docs/podman-pod-start.1.md]: sed s/containers/pods

File: docs/podman-pod-stop.1.md [hunk 20]
```diff
@@ -0,0 +1,53 @@
+% podman-pod-stop "1"
+
+## NAME
+podman\-pod\-stop - Stop one or more containers
```
Comment (rhatdan) [inline docs/podman-pod-stop.1.md]: s/containers/pods/g

File: docs/podman-pod-stop.1.md [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+## DESCRIPTION
+Stop containers in one or more pods.  You may use pod IDs or names as input.
+
+## OPTIONS
+
+**--all, -a**
+
+Stops all pods
+
+**--cleanup**
```
Comment (rhatdan) [inline docs/podman-pod-stop.1.md]: Why?

**Comment by rhatdan** [comment]
I might take away "else" from your laptop.  :^)

File: cmd/podman/pod_start.go [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+		Flags:                  podStartFlags,
+		Action:                 podStartCmd,
+		ArgsUsage:              "POD-NAME [POD-NAME ...]",
+		UseShortOptionHandling: true,
+	}
+)
+
+func podStartCmd(c *cli.Context) error {
+	args := c.Args()
+	if (c.Bool("all") || c.Bool("latest")) && len(args) > 0 {
```
Comment (mheon) [inline cmd/podman/pod_start.go]: Can you make a helper function to do this? It seems to come up often... Just passing in the cli.Context and checking for good options seems fine.

File: cmd/podman/pod_start.go [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+				}
+				lastError = errors.Wrapf(err, "unable to find pod %s", i)
+				continue
+			}
+			pods = append(pods, pod)
+		}
+	}
+
+	for _, pod := range pods {
+		ctr_errs, err := pod.Start(getContext())
```
Comment (mheon) [inline cmd/podman/pod_start.go]: Should save getContext() somewhere and pass it in to every call here

File: cmd/podman/pod_start.go [hunk 24]
```diff
@@ -0,0 +1,105 @@
...
+		if err != nil {
+			if lastError != nil {
+				fmt.Fprintln(os.Stderr, lastError)
+			}
+			lastError = errors.Wrapf(err, "unable to start pod %q", pod.ID())
+			continue
+		} else if ctr_errs != nil {
+			for ctr, err := range ctr_errs {
+				if lastError != nil {
+					fmt.Fprintln(os.Stderr, lastError)
```
Comment (mheon) [inline cmd/podman/pod_start.go]: Not logrus.Errorf?

File: cmd/podman/pod_start.go [hunk 25]
```diff
@@ -0,0 +1,105 @@
...
+			}
+			pods = append(pods, pod)
+		}
+	}
+
+	for _, pod := range pods {
+		ctr_errs, err := pod.Start(getContext())
+		if err != nil {
+			if lastError != nil {
+				fmt.Fprintln(os.Stderr, lastError)
```
Comment (mheon) [inline cmd/podman/pod_start.go]: Not logrus.Errorf?

File: cmd/podman/pod_start.go [hunk 26]
```diff
@@ -0,0 +1,105 @@
...
+		}
+		pods = append(pods, pod)
+	} else {
+		for _, i := range args {
+			pod, err := runtime.LookupPod(i)
+			if err != nil {
+				if lastError != nil {
+					fmt.Fprintln(os.Stderr, lastError)
+				}
+				lastError = errors.Wrapf(err, "unable to find pod %s", i)
```
Comment (mheon) [inline cmd/podman/pod_start.go]: Not logrus.Errorf?

File: cmd/podman/pod_rm.go (hunk 4 above)
Comment (haircommander) [reply cmd/podman/pod_rm.go]: yeah that's true, though I kind of feel having an if, then if, then else looks clunky:
```
    if c.Bool("all") {                                                                                                                                                                                             
        pods, err = runtime.Pods()                                                                                                                                                                                 
        if err != nil {                                                                                                                                                                                            
            return errors.Wrapf(err, "unable to get running pods")                                                                                                                                                 
        }                                                                                                                                                                                                          
    }                                                                                                                                                                                                              
    if c.Bool("latest") {                                                                                                                                                                                          
        pod, err := runtime.GetLatestPod()                                                                                                                                                                         
        if err != nil {                                                                                                                                                                                            
            return errors.Wrapf(err, "unable to get latest pod")                                                                                                                                                   
        }                                                                                                                                                                                                          
        pods = append(pods, pod)                                                                                                                                                                                   
    } else {                                                                                                                                                                                                       
        for _, i := range args {                                                                                                                                                                                   
            pod, err := runtime.LookupPod(i)                                                                                                                                                                       
            if err != nil {                                                                                                                                                                                        
                if lastError != nil {                                                                                                                                                                              
                    fmt.Fprintln(os.Stderr, lastError)                                                                                                                                                             
                }                                                                                                                                                                                                  
                lastError = errors.Wrapf(err, "unable to find pod %s", i)                                                                                                                                          
                continue                                                                                                                                                                                           
            }                                                                                                                                                                                                      
            pods = append(pods, pod)                                                                                                                                                                               
        }                                                                                                                                                                                                          
    }  
```
up to you.

File: cmd/podman/pod_stop.go (hunk 11 above)
Comment (haircommander) [reply cmd/podman/pod_stop.go]: Not sure, I didn't really understand what it did. I can remove the option if that's preferred.

File: cmd/podman/pod_rm.go (hunk 4 above)
Comment (rhatdan) [reply cmd/podman/pod_rm.go]: I am saying you need no elses here at all.
You are appending the pods on .

File: cmd/podman/pod_stop.go (hunk 11 above)
Comment (rhatdan) [reply cmd/podman/pod_stop.go]: Well if you don't know what cleanup option is for, then remove it.

File: docs/podman-pod-stop.1.md (hunk 21 above)
Comment (haircommander) [reply docs/podman-pod-stop.1.md]: I'm not sure!

Comment (rhatdan) [reply docs/podman-pod-stop.1.md]: Kill it.

File: cmd/podman/pod_stop.go (hunk 11 above)
Comment (mheon) [reply cmd/podman/pod_stop.go]: Hardcode it to true for podman

File: cmd/podman/pod_rm.go (hunk 4 above)
Comment (haircommander) [reply cmd/podman/pod_rm.go]: ...long day
got it

File: docs/podman-pod-stop.1.md (hunk 21 above)
Comment (haircommander) [reply docs/podman-pod-stop.1.md]: for clarity: kill it is in kill this option, or kill it as in that is the reason why one would use --cleanup?

**Comment by rhatdan** [comment]
bot, retest this please

**Comment by mheon** [comment]
Oh dear, the tests are actually green.

**Comment by mheon** [comment]
At some point we should consider allowing the IDs of individual containers that were stopped/started to be printed during the output, maybe like so:

podman start abcde
abced
- ctr1id
- ctr2id
- ctr3id

But that would require extra work on the libpod side to expose the state changes that happened.

LGTM, this is a good start.

**Comment by haircommander** [comment]
SGTM, the change would be effectively trivial on the podman side. I wonder if a -q should supress it, and have it print ctr ids by default?

**Comment by mheon** [comment]
I think we'd have a `-q` which would print just pod IDs, and maybe a `-v` for verbose that would print container IDs as well? We could even get fancy and add a `-vv` for double-verbose that would print the IDs of both containers that started and containers that were already started.

**Comment by haircommander** [comment]
@mheon then what would the default be? I feel like:
-q for pod ID,
nothing for pod and ctr-id
-v for pod and ctr-id along with those already started
if the latter would be useful.

**Comment by mheon** [comment]
That sounds like a good way of doing it, yeah

**Comment by haircommander** [comment]
I'll focus on the rest of the pod commands first, but I can look into that after.

Speaking of @rhatdan PTAL it'd be cool to get this in, it'll make testing the next commands easier

**Comment by mheon** [comment]
I'm going to merge this so we can cut a release. Namespaces will wait until Monday.

**Comment by haircommander** [comment]
I'd actually like to add name spaces to pod create and ps anyway so I think it makes sense to wait

**Comment by mheon** [comment]
Should just work on your side - all the magic happens in the database.

**Comment by haircommander** [comment]
at least with ps I need to manually include it in the output.

**Comment by mheon** [comment]
I don't think it's worth including in `ps` output - seems more like an `inspect` thing to me.

**Comment by haircommander** [comment]
crictl includes it in their pods output, which is our ps. so I think @rhatdan wants it in for CRI-O

**Pull request merged**